athena-elasticsearch/src/main/java/com/amazonaws/connectors/athena/elasticsearch/ElasticsearchRecordHandler.java

```diff
@@ -128,27 +135,36 @@ protected void readWithConstraint(BlockSpiller spiller, ReadRecordsRequest recor
 
         String domain = recordsRequest.getTableName().getSchemaName();
         String endpoint = recordsRequest.getSplit().getProperty(domain);
+        String shard = recordsRequest.getSplit().getProperty(ElasticsearchMetadataHandler.SHARD_KEY);
         long numRows = 0;
 
         if (queryStatusChecker.isQueryRunning()) {
-            AwsRestHighLevelClient client = clientFactory.getClient(endpoint);
+            AwsRestHighLevelClient client = clientFactory.getOrCreateClient(endpoint);
             try {
                 // Create field extractors for all data types in the schema.
                 GeneratedRowWriter rowWriter = createFieldExtractors(recordsRequest);
 
                 // Create a new search-source injected with the projection, predicate, and the pagination batch size.
                 SearchSourceBuilder searchSource = new SearchSourceBuilder().size(QUERY_BATCH_SIZE)
+                        .timeout(searchDocumentTimeout)
                         .fetchSource(ElasticsearchQueryUtils.getProjection(recordsRequest.getSchema()))
                         .query(ElasticsearchQueryUtils.getQuery(recordsRequest.getConstraints().getSummary()));
                 // Create a new search-request for the specified index.
-                SearchRequest searchRequest = new SearchRequest(recordsRequest.getTableName().getTableName());
+                SearchRequest searchRequest = new SearchRequest(recordsRequest.getTableName().getTableName())
+                        .preference(shard);
                 int hitsNum;
                 int currPosition = 0;
                 do {
                     // Process the search request injecting the search-source, and setting the from position
                     // used for pagination of results.
                     SearchResponse searchResponse = client
                             .getDocuments(searchRequest.source(searchSource.from(currPosition)));
+
+                    // Throw on query timeout.
+                    if (searchResponse.isTimedOut()) {
+                        throw new IOException("Request for " + shard + " timed out.");
+                    }
```

Comment:
im not sure this should be an IOException, also when does this happen?

Reply:
Not sure we need a timeout check here - removed.

Reply:
Doesn't look like a timeout check is needed here - removed.